Improved operator<=> slides (#236)
hsf-training/cpluspluscourse: talk/expert/cpp20spaceship.tex

\subsection[spaceship]{The \texttt{<=>} operator}

\begin{frame}[fragile]
  \frametitle{The burden of comparison operators}
  \begin{block}{Motivation}
    \begin{itemize}
    \item One often needs \mintinline{cpp}{operator<} for a home-made class
      \begin{itemize}

>       \item e.g. when sorting a \mintinline{cpp}{std::vector} of it

Reviewer: ```suggestion
      \item e.g. when sorting a \mintinline{cpp}{std::vector}
```

>       \item e.g. when sorting a \mintinline{cpp}{std::vector} of it
>       \item e.g. when using it as a key for \mintinline{cpp}{std::set} or \mintinline{cpp}{std::map}

Reviewer: This decreases the extra space after the dot
```suggestion
      \item e.g.\ when sorting a \mintinline{cpp}{std::vector} of it
      \item e.g.\ when using it as a key for \mintinline{cpp}{std::set} or \mintinline{cpp}{std::map}
```

Author: Does not really seem to change anything, and the current spacing is fine on top. Which problem did you try to solve ?

      \end{itemize}
    \item Many other operators are also desirable for completeness
      \begin{itemize}
      \item \mintinline{cpp}{operator>}, \mintinline{cpp}{operator>=}, \mintinline{cpp}{operator<=}, ...

>       \item often implemented reusing e.g. \mintinline{cpp}{operator<} and \mintinline{cpp}{operator==}

Reviewer: ```suggestion
      \item often implemented reusing e.g.\ \mintinline{cpp}{operator<} and \mintinline{cpp}{operator==}
```

Author: This looked also fine. So same question : what was the problem ?

  \frametitlecpp[20]{The three-way comparison operator}
  \begin{block}{Idea}
    \begin{itemize}
    \item C++20 introduces \mintinline{cpp}{operator<=>}
    \item named the \textbf{three-way comparison operator}.
      \begin{itemize}
      \item inofficially called \textbf{spaceship operator}
      \end{itemize}
    \item allowing to implement all comparisons in one go
    \end{itemize}
  \end{block}
  \begin{exampleblock}{How it works}
    \begin{itemize}

>     \item if returns \emph{something} which can be compared to \mintinline{cpp}{0}

Reviewer: ```suggestion
    \item it returns \emph{something} which can be compared to \mintinline{cpp}{0}
```

    \begin{cppcode*}{}
    template <typename T>
    void three_way_compare( T lhs, T rhs ) {
      auto res = ( lhs <=> rhs );
      std::cout << "(" << lhs << "<=>" << rhs << ") : "
                << (res<0) << (res==0) << (res>0)
                << std::endl;
    }
    int main() {
      three_way_compare(1, 2); // (1<=>2): 100
      three_way_compare(2, 2); // (2<=>2): 010
      three_way_compare(2, 1); // (2<=>1): 001
    }
    \end{cppcode*}
  \end{exampleblock}
\end{frame}

\begin{frame}[fragile]
  \frametitlecpp[20]{Different kinds of ordering}
  \begin{block}{About the return type of \texttt{operator<=>}}
    \begin{itemize}
    \item for integers \mintinline{cpp}{operator<=>} returns \mintinline{cpp}{std::strong_ordering}
    \item \mintinline{cpp}{partial_ordering} and \mintinline{cpp}{weak_ordering} also exist
    \end{itemize}
  \end{block}
  \begin{exampleblock}{3 types of ordering}
    \begin{description}[partial]
    \item[strong] exactly one test among \mintinline{cpp}{<0}, \mintinline{cpp}{==0}, and \mintinline{cpp}{>0} will return \mintinline{cpp}{true}
    \item[partial] like strong but all tests may return \mintinline{cpp}{false}
      \begin{itemize}

>       \item e.g. floating points for \mintinline{cpp}{NaN}

Reviewer: ```suggestion
      \item e.g.\ compare a floating point number to \mintinline{cpp}{NaN}
```

      \end{itemize}
    \item[weak] like strong but two \emph{equivalent} values may differ
      \begin{itemize}
      \item they are however \emph{equivalent} for ranking

>       \item e.g. rationals $2/3$ and $4/6$

Reviewer: ```suggestion
      \item e.g.\ rational numbers $2/3$ and $4/6$
```

      \end{itemize}
    \end{description}
  \end{exampleblock}
\end{frame}

\begin{frame}[fragile]
  \frametitlecpp[20]{Exercising different orderings}
  \scriptsize
  \begin{exampleblock}{Example}
    \begin{cppcode*}{}
    struct Ratio {
      unsigned n, d ;
      friend std::weak_ordering operator<=>( Ratio const & a,
                                             Ratio const & b ) {
        return (a.n * b.d) <=> (a.d * b.n);
      }
      friend std::ostream & operator<<( std::ostream & os, Ratio const & r ) {
        return (os << r.n << '/' << r.d);
      }
    };
    int main() {
      // floats use partial_ordering
      three_way_compare(+0., -0.);     // (0<=>-0)   : 010
      three_way_compare(0./1., 1./0.); // (0<=>inf)  : 100
      three_way_compare(0., 0./0.);    // (0<=>-nan) : 000

      // Ratio uses weak_ordering
      three_way_compare(Ratio{1, 2}, Ratio{2, 3}); // (1/2<=>2/3) : 100
      three_way_compare(Ratio{2, 3}, Ratio{4, 6}); // (2/3<=>4/6) : 010
      three_way_compare(Ratio{2, 3}, Ratio{1, 2}); // (2/3<=>1/2) : 001
    }
    \end{cppcode*}
  \end{exampleblock}
\end{frame}

\begin{frame}[fragile]

>   \frametitlecpp[20]{Compiler generated comparison operators}

Reviewer: ```suggestion
  \frametitlecpp[20]{Compiler-generated comparison operators}
```

  \begin{block}{For a given home-made class}
    \begin{itemize}
    \item defining \mintinline{cpp}{operator<=>} allows the compiler to generate \mintinline{cpp}{operator<}, \mintinline{cpp}{operator>}, \mintinline{cpp}{operator<=} and \mintinline{cpp}{operator=>}

>     \item of course, one can overwrite the default implementations

Reviewer: Not sure if this word is better. Feel free to disregard this.
```suggestion
    \item of course, one can replace the default implementations
```

Reviewer 2: In a previous debate, we emphasized  that the compiler does not really generate < (and > and...) in the class. Rather, each time it encounters a use of < in the code, it will rephrase it in terms of <=> (TO BE CONFIRMED). This can be seen as a detail, but perhaps it may help to known this when trying to decipher a compilation error involving those operators.

Author: @chavid, I've replaced the generate line with "allows the compiler to use it when encountering other comparison operators", so that we stay vague enough

    \item the compiler will \emph{NOT} add a default implementation for \mintinline{cpp}{operator==} and \mintinline{cpp}{operator!=}
      \begin{itemize}
      \item as those operators mean \textbf{equal}, rather than \textbf{equivalent}
      \item if \mintinline{cpp}{operator<=>} does not provide a strong order, it is advised not to define \mintinline{cpp}{operator==} yourself
      \end{itemize}
    \end{itemize}
  \end{block}
\end{frame}

\begin{frame}[fragile]

>   \frametitlecpp[20]{Compiler generated \texttt{operator<=>}}

Reviewer: ```suggestion
  \frametitlecpp[20]{Compiler-generated \texttt{operator<=>}}
```

  \begin{block}{Default \texttt{operator<=>} implementation}
    \begin{itemize}

>     \item One can ask the compiler to provide a default implementation for \mintinline{cpp}{operator<=>} and/or \mintinline{cpp}{opeartor==}

Reviewer: ```suggestion
    \item One can ask the compiler to provide a default implementation for \mintinline{cpp}{operator<=>} and/or \mintinline{cpp}{operator==}
```

      \begin{itemize}
        \item declaring them with ``\mintinline{cpp}{= default}''
        \item it will compare the member variables, one by one
      \end{itemize}

>     \item Can be useful e.g. for tuples
>     \item Can be wrong, e.g. for the Ratio class

Reviewer: ```suggestion
    \item Can be useful e.g.\ for tuples
    \item Can be wrong, e.g.\ for the Ratio class
```

Author: ditto

    \item If \mintinline{cpp}{operator<=>} is defaulted and no \mintinline{cpp}{operator==} is defined, then the compiler also provides \mintinline{cpp}{operator==}
    \end{itemize}
  \end{block}
\end{frame}

\begin{frame}[fragile]
  \frametitlecpp[20]{\texttt{operator<=>} summary}
  \begin{alertblock}{Summary}
    \begin{itemize}

>       \item Defining \mintinline{cpp}{operator<=>} allows you to get \mintinline{cpp}{operator<}, \mintinline{cpp}{operator>},\mintinline{cpp}{operator<=}, and \mintinline{cpp}{operator>=} for free

Reviewer 2: I originally deliberately suggested `allows you to use`, rather than `allow you to get`, so to not let think the reader that the operators are generated.

Author: indeed, you're fully right. Fixed
